modules/imap/functions.php

```diff
@@ -446,11 +448,33 @@ function format_msg_part_row($id, $vals, $output_mod, $level, $part, $dl_args, $
                 '</td><td class="part_charset">'.(isset($vals['attributes']['charset']) && trim($vals['attributes']['charset']) ? $output_mod->html_safe(strtolower($vals['attributes']['charset'])) : '');
         }
         $res .= '</td><td class="part_desc">'.$output_mod->html_safe(decode_fld($desc)).'</td>';
-        $res .= '<td class="download_link"><a href="?'.$dl_args.'&amp;imap_msg_part='.$output_mod->html_safe($id).'">'.$output_mod->trans('Download').'</a></td></tr>';
+        $res .= '<td class="download_link"><a href="?'.$dl_args.'&amp;imap_msg_part='.$output_mod->html_safe($id).'">'.$output_mod->trans('Download').'</a></td>';
+        if ($lc_type == "textplain" || $lc_type == "multipartmixed") {
+            $res .= '<td class="download_link"><a href="?'.$dl_args.'&amp;imap_msg_part='.$output_mod->html_safe($id).'"></a></td>';
+        }
+    }
+    if (isset($vals['file_attributes']['attachment'])) {
+        $res .= '<td><a href="?'.$at_args.'&amp;imap_msg_part='.$output_mod->html_safe($id).'" class="enable_remove">'.$output_mod->trans('Remove').'</a></td>';
     }
+    $res .= '</tr>';
     return $res;
 }}
 
+/*
+ * Returns the modified message
+ * @param string $attachment_name filename
+ * @param string $msg raw message
+ * @return string
+ */
+if (!hm_exists('remove_attachment')) {
+    function remove_attachment($msg_id, $msg) {
+        $message = Message::from($msg, false);
+        $message->removeAttachmentPart($msg_id - 2);
```

Comment:
Are you sure it is always 2 here? It seems to depend on the actual mime message structure which might have different parts before the attachments - text only, text and html, html-only, mutlipart with several parts before the attachment.

Reply:
@josaphatim I think we need a specific part number here, not an arbitrary number at the end, don't we?

Reply:
You are right @kroky. But I don't get a way to use a fixed index. When I take a look at https://github.com/zbateson/mail-mime-parser getAllAttachmentParts function it returns a simple array with keys starting from 0. While attachments in cypht start from 2. That why I did $msg_id - 2 for them to correspond.

Reply:
Did you try a multipart message with several text parts (more than 2)?

Reply:
Now found the pattern. https://github.com/zbateson/mail-mime-parser attachments combines embedded images and file attachments. It uses 0-based index. So I created a function to get corresponding index from the given message part. Also I added a confirmation before deleting attachment.

Reply:
Yes, that's much better, thanks!